### PR TITLE
editor: snap to line start/end when vertical nav would be a no-op

### PR DESCRIPTION
**Pull request by kotachisam**
Up on the first line and Down on the last line previously did nothing. Snap to Head/End respectively so the keypress always has a useful effect, mirroring the double-Esc convention of turning a dead key into something helpful. Modifier keys (shift/ctrl/alt) bypass the snap so selection extension and other bindings continue to work as before.

Applies to default-navigation paths in insert and normal only; vim-style jk bindings in normal mode have now been adjusted to match with the Up/Down arrow keys (commit [73dd456](https://github.com/AmmarAbouZor/tui-journal/pull/628/commits/73dd456081a0304745d917d63ee4fd5a315fdf9e). Visual mode is left alone: plain arrows there already cancel selection, and we don't want to introduce a competing snap behavior. 

## Manual testing
 - Insert mode, single-line entry: Up/Down snap to start/end
 - Multi-line, cursor on first line: Up snaps to Head; mid-buffer still moves up
 - Multi-line, cursor on last line: Down snaps to End
 - Shift+Up on first line: still extends selection (no snap) 
 - Visual mode: unaffected — arrow keys keep their existing cancel-selection behavior
 - Normal mode k on first line: snaps to Head.
 - Normal mode j on last line, snaps to End.

## Contribution and AI Policy
- [x] I have read and accept the Contribution and AI Policy in `README.md`, and I have reviewed this PR myself.

### CODE REVIEW

**Comment by AmmarAbouZor** [comment]
Thanks for the suggestions. I've never noticed this feature or used it before in other editors as it's not used in neovim.

We can have it in, but we need to have the same behavior for `j` and `k` in normal mode since the two keys also move up and down one line

**Comment by kotachisam** [comment]
I will add these today, good point well-made. I fluctuate between vim and nano, have yet to graduate to neovim!

**Comment by kotachisam** [comment]
Apologies for the delay Ammar, have made the changes so that in insert mode, the directional buttons have snap behaviour, and in normal mode, both the directional buttons and j/k exhibit that behaviour.

File: src/app/ui/editor/mod.rs [hunk 1]
```diff
@@ -25,6 +25,12 @@ pub enum EditorMode {
     Visual,
 }
 
+#[derive(Debug, Clone, Copy)]
+enum VerticalDir {
+    Up,
+    Down,
+}
+
```
Comment (AmmarAbouZor) [inline src/app/ui/editor/mod.rs]: We should reorder the place of this struct in file to match following:
- pub types (Here `Editor`)
- private struct like this one
- impl block for pub types
- impl block for private types

--------------

Also it's useful to have this type derive `PartialEq` and `Eq`

---------

Finally,  a one line docs for the type (not the fields) would be nice

**Review by AmmarAbouZor — CHANGES_REQUESTED** [review]
I think, we should take over the whole up & down cursor movement instead check and hijack it's behavior. This will include:
* Up and Down aren't default navigation anymore.
* The entire handling for Up and Down is handled by us; It should check if it needs to snap and apply it. Otherwise it should move the cursor itself and not handling it back to the old logic.
* Also it should handle cases where users are selecting text using shift in insert mode and vim visual (Copy) mode which is missing currently.

At the end I would like to have function like: move_cursor_up() that will take care of all of the cases in one place without jumping around between old logic and new one.

----
I think also that the required changes are getting more complicated, so please feel free to ask questions or to hand this feature over if you feel it's getting too complicated for your initial implementation concepts

**Comment by kotachisam** [comment]
Hey Ammar, sorry for the delay in getting back to this. Thanks for the guidance - the "take it over completely" framing made this a lot cleaner to reason about.

Up/Down are now pulled into a single `move_cursor_vertical `(with `move_cursor_up`/`move_cursor_down` on top), so they're no longer default-navigation and never fall back to the old path. It owns the snap-or-move decision and the selection state in one place. Shift+arrows extend a selection in insert mode, and visual mode extends on both the arrows and j/k. Also did the VerticalDir tidy-up you flagged (moved below Editor, derives PartialEq/Eq, one-line doc), and added a small unit-test module since there wasn't any editor coverage yet.

Two things I'd genuinely value your vim intuition on - I bounce between vim and nano, and I definitely don't use vim to its full potential, so I'd rather not guess wrong for people who actually live in vim:

1. At the very top/bottom line while a selection is active, it currently snaps the cursor to line start/end (so the selection grabs to the buffer edge). Would you expect it to just stop there instead - the way k on line 1 does nothing in vim? I can see benefits of both ways.
2. Should Shift+Up/Down start a selection in normal mode too, or is that un-vim and normal mode should stay pure navigation? I left it navigation-only for now, but in my head I would've made it start a selection in normal mode. I wasn't sure that would match how you'd reach for it.

Let me know your thoughts, and I can push further changes as required.

**Comment by AmmarAbouZor** [comment]
Sorry for the late replay. I missed the notification of your replay and noticed it today!

Here are the answers for the remaining points.

1. I would go with the more simple implementation here, as both of options are good.
2. Current implementation starts selection with shift + arrows in normal mode and it stays in that mode, without switching modes. The idea there is to not force vim special modes on users who aren't familiar with that so I would keep it like this. I would expect snap to line start/end to work with selection with shift as a normal editor (Like the comments editor in GitHub UI) to work.

I didn't give it another review as it's still not done. Please let me know once all changes from your side are implemented to give it a final review

**Comment by kotachisam** [comment]
Rebased onto current main and pushed both changes.

On point 2 - good catch. Pulling Up/Down out of `is_default_navigation` meant normal mode stopped extending a selection on Shift+Arrow, which main did via `move_cursor_with_shift`. `should_extend_selection` now keys off Shift in any mode (visual still always extends), so normal mode selects with Shift, stays in normal mode, and snaps to line start/end while doing it, which is the GitHub-comment-box behaviour you described. 

On point 1 - kept the simpler unified path, so snapping applies whether or not a selection is active. That falls out of the same change rather than needing a special case.

One extra fix while I was in there: vertical movement no longer touches selection state when the cursor can't actually move (already at line start on the first line, or line end on the last). It was leaving a zero-width selection behind, so a subsequent copy returned an empty string - and with `sync_os_clipboard = true` that silently cleared the OS clipboard. `tui-textarea` avoids this by only mutating selection inside `if let Some(cursor) = next_cursor(...); resolve_vertical_move` now does the same.

Added three unit tests covering normal-mode shift selection and both dead-edge cases. `fmt, clippy -D warnings` and the full suite pass. Ready for your final review.

File: src/app/ui/editor/mod.rs [hunk 2]
```diff
@@ -527,14 +606,26 @@ impl<'a> Editor<'a> {
...
 fn is_default_navigation(input: &Input) -> bool {
     let has_control = input.modifiers.contains(KeyModifiers::CONTROL);
     let has_alt = input.modifiers.contains(KeyModifiers::ALT);
     match input.key_code {
         KeyCode::Left
         | KeyCode::Right
-        | KeyCode::Up
-        | KeyCode::Down
```
Comment (AmmarAbouZor) [inline src/app/ui/editor/mod.rs:537]: We shouldn't remove the keycodes here to preserve the behavior when Up and Down are pressed with a modifier other than shift (Like control).
The current implementation for `Ctrl+Up` for example will skip `vertical_arrow_dir()` and also skip `is_default_navigation()` moving to vim navigation function which isn't correct

File: src/app/ui/editor/mod.rs [hunk 3]
```diff
@@ -527,14 +606,26 @@ impl<'a> Editor<'a> {
     }
 }
 
+/// Maps a plain or Shift-modified Up/Down arrow to a vertical direction so that
+/// vertical movement is fully owned by [`Editor::move_cursor_vertical`]. Arrows
+/// carrying Ctrl/Alt keep their default-navigation handling.
+fn vertical_arrow_dir(input: &Input) -> Option<VerticalDir> {
+    if !(input.modifiers - KeyModifiers::SHIFT).is_empty() {
+        return None;
+    }
```
Comment (AmmarAbouZor) [inline src/app/ui/editor/mod.rs]: I like more simple code in such places making the readability much simpler for my future self

```suggestion
   let modifiers = input.modifiers;
   if !modifiers.is_empty() && modifiers != KeyModifiers::SHIFT {
       return None;
   }
```

**Review by AmmarAbouZor — CHANGES_REQUESTED** [review]
Thanks for the changes. The code looks clean now but I've found a regression regarding using arrows with modifier other than shift.

Also I've added other comment regarding making the code less smart but more understandable.

#### Note:
Current implementation doesn't handle the very edge case for Up&Down with Ctrl modifier in case the cursor in already in first/end line. 
However, I don't want to include handling it as it will make the code much more complex to cover an edge case that no one will ever think about.

File: src/app/ui/editor/mod.rs (hunk 2 above)
Comment (kotachisam) [reply src/app/ui/editor/mod.rs:537]: Restored. Confirmed the fall-through you described: Ctrl+Up gets None from `vertical_arrow_dir`, then fails `is_default_navigation`, and lands in `handle_vim_motions`. Ordering is still safe because `vertical_arrow_dir` runs first in both paths, so plain/Shift arrows keep the snap. Added a test asserting Ctrl/Alt + Up/Down stay on default navigation.

File: src/app/ui/editor/mod.rs (hunk 3 above)
Comment (kotachisam) [reply src/app/ui/editor/mod.rs]: Applied. Behaviour is identical for Ctrl+Shift either way, so it's a free readability win.

**Comment by kotachisam** [comment]
> #### Note:
> Current implementation doesn't handle the very edge case for Up&Down with Ctrl modifier in case the cursor in already in first/end line. However, I don't want to include handling it as it will make the code much more complex to cover an edge case that no one will ever think about.

Both addressed. Restored the keycodes and took your version of the modifier check. Agree on leaving the Ctrl+Up/Down first/last-line case alone - not worth the complexity.

**Review by AmmarAbouZor — APPROVED** [review]
Looks great. Thanks for the implementation in the fixes 🙏
